server/src/main/java/io/druid/server/namespace/cache/LookupCoordinatorManager.java

```diff
@@ -0,0 +1,571 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ *
+ */
+public class LookupCoordinatorManager
+{
+  public static final String LOOKUP_CONFIG_KEY = "lookups";
+  // Doesn't have to be the same, but it makes things easy to look at
+  public static final String LOOKUP_LISTEN_ANNOUNCE_KEY = LOOKUP_CONFIG_KEY;
+  private static final Logger LOG = new Logger(LookupCoordinatorManager.class);
+
+  private final ListeningScheduledExecutorService executorService;
+  private final ListenerDiscoverer listenerDiscoverer;
+  private final HttpClient httpClient;
+  private final ObjectMapper smileMapper;
+  private final JacksonConfigManager configManager;
+  private final Object startStopSync = new Object();
+  // Updated by config watching service
+  private AtomicReference<Map<String, Map<String, Map<String, Object>>>> lookupMapConfigRef;
+  private volatile Map<String, Map<String, Map<String, Object>>> prior_update = ImmutableMap.of();
+  private volatile boolean started = false;
+
+  private final static Function<HostAndPort, URL> HOST_TO_URL = new Function<HostAndPort, URL>()
+  {
+    @Nullable
+    @Override
+    public URL apply(HostAndPort input)
+    {
+      if (input == null) {
+        LOG.warn("null entry in namespaces");
+        return null;
+      }
+      try {
+        return getLookupsURL(input);
+      }
+      catch (MalformedURLException e) {
+        LOG.warn(e, "Skipping node. Malformed URL from `%s`", input);
+        return null;
+      }
+    }
+  };
+
+  @Inject
+  public LookupCoordinatorManager(
+      final @Global HttpClient httpClient,
+      final ListenerDiscoverer listenerDiscoverer,
+      final @Smile ObjectMapper smileMapper,
+      final JacksonConfigManager configManager
+  )
+  {
+    this.listenerDiscoverer = listenerDiscoverer;
+    this.configManager = configManager;
+    this.httpClient = httpClient;
+    this.smileMapper = smileMapper;
+    executorService = MoreExecutors.listeningDecorator(
+        Executors.newScheduledThreadPool(10, Execs.makeThreadFactory("LookupCoordinatorManager--%s"))
+    );
+  }
+
+  void deleteOnHost(final URL url)
+      throws ExecutionException, InterruptedException, IOException
+  {
+    final AtomicInteger returnCode = new AtomicInteger(0);
+    final AtomicReference<String> reasonString = new AtomicReference<>(null);
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Dropping %s", url);
+    }
+
+    try (final InputStream result = httpClient.go(
+        new Request(HttpMethod.DELETE, url)
+            .addHeader(HttpHeaders.Names.ACCEPT, SmileMediaTypes.APPLICATION_JACKSON_SMILE),
+        makeResponseHandler(returnCode, reasonString),
+        Duration.millis(1_000) // TODO: configurable
```

Comment:
Yes, any remaining TODOs should be relatively minor compared to the overall logic of the PR though.